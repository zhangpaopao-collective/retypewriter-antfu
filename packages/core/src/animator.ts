import { calculatePatches, diff } from './patch'
import type { Patch } from './types'

export interface AnimatorStep {
  index: number
  char?: string
  cursor: number
  content: string
}

export function* createAnimator(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let _input = input
  for (let patchIndex = 0; patchIndex < patches.length; patchIndex++) {
    const patch = patches[patchIndex]
    if (patch.type === 'insert') {
      for (let i = 0; i < patch.text.length; i++) {
        const index = patch.from + i
        _input = _input.slice(0, index) + patch.text[i] + _input.slice(index)
        yield {
          index: patchIndex,
          char: patch.text[i],
          cursor: index,
          content: _input,
        }
      }
    }
    else {
      for (let i = 0; i < patch.length; i++) {
        const index = patch.from - i
        _input = _input.slice(0, index - 1) + _input.slice(index)
        yield {
          index: patchIndex,
          cursor: index,
          content: _input,
        }
      }
    }
  }
}

export function simpleAnimator(input: string, output: string) {
  const delta = diff(input, output)
  const patches = calculatePatches(delta)

  return createAnimator(input, patches)
}
