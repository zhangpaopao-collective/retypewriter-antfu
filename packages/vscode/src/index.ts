import { commands, window } from 'vscode'

export interface Snapshot {
  content: string
  time: number
}

export function activate() {
  window.showInformationMessage('Hi')

  const snapMap = new Map<string, Snapshot[]>()

  commands.registerCommand('retypewriter.snap', () => {
    const doc = window.activeTextEditor?.document
    if (!doc)
      return
    const path = doc.uri.fsPath
    if (!snapMap.has(path))
      snapMap.set(path, [])
    snapMap.get(path)!.push({
      content: doc.getText(),
      time: Date.now(),
    })
  })

  commands.registerCommand('retypewriter.play', () => {
    const doc = window.activeTextEditor?.document
    if (!doc)
      return
    const path = doc.uri.fsPath
    const snaps = snapMap.get(path)
    if (!snaps?.length)
      window.showErrorMessage('No snapshots found')
      // return

    // const lastSnap = snaps[snaps.length - 1]
  })
}

export function deactivate() {

}
