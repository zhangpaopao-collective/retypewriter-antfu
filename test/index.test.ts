import { describe, expect, it } from 'vitest'
import { input, output } from '../examples'
import { applyPatches, calculatePatches, diff } from '../src'

describe('should', () => {
  it('diff', () => {
    const delta = diff(input, output)
    expect(delta).toMatchSnapshot('delta')

    const patches = calculatePatches(delta)
    expect(patches).toMatchSnapshot('patches')

    const applied = applyPatches(input, patches)
    expect(applied).toMatchSnapshot('applied')

    expect(output).toEqual(applied)
  })
})
