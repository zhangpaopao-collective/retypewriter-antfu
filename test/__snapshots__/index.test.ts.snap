// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should > diff > applied 1`] = `
"
  patches.push({
    type: 'removal',
    from: index,
    length: change[1].length,
  })
"
`;

exports[`should > diff > delta 1`] = `
[
  [
    0,
    "
  patches.push({
    type: '",
  ],
  [
    -1,
    "insert",
  ],
  [
    1,
    "removal",
  ],
  [
    0,
    "',
    from: index,
    ",
  ],
  [
    -1,
    "text",
  ],
  [
    1,
    "length",
  ],
  [
    0,
    ": change[1]",
  ],
  [
    1,
    ".length",
  ],
  [
    0,
    ",
  })
",
  ],
]
`;

exports[`should > diff > patches 1`] = `
[
  {
    "from": 35,
    "length": 6,
    "type": "removal",
  },
  {
    "from": 29,
    "text": "removal",
    "type": "insert",
  },
  {
    "from": 64,
    "length": 4,
    "type": "removal",
  },
  {
    "from": 60,
    "text": "length",
    "type": "insert",
  },
  {
    "from": 77,
    "text": ".length",
    "type": "insert",
  },
]
`;
